File: jpp/4/3/DHSetup.rb
# typed: true
# przy uzyciu Sorbet (type checker), aby byly typy generyczne
require 'sorbet-runtime'

class DHSetup

   
  extend T::Sig # zeby miec dostep do type_member
  extend T::Generic #  typ generyczny
  Elem = type_member # klasa DHSetup jest generyczna 
  
  def initialize
    @generator = generateGenerator
  end

  sig { returns(T) }
  def getGenerator 
    @generator 
  end 

  sig { params(a: T, b: Integer).returns(T) }
  def power(a, b)
    if b == 0 
      return T.new(1)
    end

    temp = power(a, b / 2)
    if b % 2 == 0
      temp * temp
    else
      a * temp * temp
    end
  end

  private

  @generator

  sig { returns(T) }
  def generateGenerator
    dum = T.new
    number = 1 + Random.rand(dum.getCharacteristics)
    T.new(number)
  end 

end

File: jpp/4/2/DH/DHSetup.java

import java.util.Random;

public class DHSetup<T extends GF> {

  private T generator;

  private T generateGenerator(T dum) {
    Random random = new Random();
    long number = random.nextLong(dum.getCharacteristic() - 1) + 1;
    return (T) new GF(number);
  }

  public DHSetup(T dum) {
    this.generator = generateGenerator(dum);
  }

  public T getGenerator() {
    return generator;
  }

  public T power(T a, long b) {
    if (b == 0) {
        return (T) new GF(1);
    }
    T temp = power(a, b/2);
    if (b % 2 == 0) {
      return (T) temp.mult(temp);
    } else {
      return (T) a.mult(temp).mult(temp);
    }
  }
}
